bcf_sr_sort.c

```diff
@@ -0,0 +1,645 @@
+
+        memmove(&srt->cnt[jvset], &srt->cnt[jvset+1], sizeof(int)*(srt->nvset - jvset - 1));
+    }
+    srt->nvset--;
+}
+int merge_vsets(sr_sort_t *srt, int ivset, int jvset)
+{
+    int i,j;
+    if ( ivset > jvset ) { i = ivset; ivset = jvset; jvset = i; }
+
+    varset_t *iv = &srt->vset[ivset];
+    varset_t *jv = &srt->vset[jvset];
+
+    kbs_bitwise_or(iv->mask,jv->mask);
+
+    i = iv->nvar;
+    iv->nvar += jv->nvar;
+    hts_expand(int, iv->nvar, iv->mvar, iv->var);
+    for (j=0; j<jv->nvar; j++,i++) iv->var[i] = jv->var[j];
+
+    int *imat = srt->pmat + ivset*srt->ngrp;
+    int *jmat = srt->pmat + jvset*srt->ngrp;
+    for (i=0; i<srt->ngrp; i++) imat[i] += jmat[i];
+    srt->cnt[ivset] += srt->cnt[jvset];
+
+    remove_vset(srt, jvset);
+
+    return ivset;
+}
+void push_vset(sr_sort_t *srt, int ivset)
+{
+    varset_t *iv = &srt->vset[ivset];
+    int i,j;
+    for (i=0; i<srt->sr->nreaders; i++)
+    {
+        vcf_buf_t *buf = &srt->vcf_buf[i];
+        buf->nrec++;
+        hts_expand(bcf1_t*,buf->nrec,buf->mrec,buf->rec);
+        buf->rec[buf->nrec-1] = NULL;
+    }
+    for (i=0; i<iv->nvar; i++)
+    {
+        var_t *var = &srt->var[ iv->var[i] ];
+        for (j=0; j<var->nvcf; j++)
+        {
+            int jvcf = var->vcf[j];
+            vcf_buf_t *buf = &srt->vcf_buf[jvcf];
+            buf->rec[buf->nrec-1] = var->rec[j];
+        }
+    }
+    remove_vset(srt, ivset);
+}
+
+static int cmpstringp(const void *p1, const void *p2)
+{
+    return strcmp(* (char * const *) p1, * (char * const *) p2);
+}
+void debug_vsets(sr_sort_t *srt)
+{
+    int i,j,k;
+    for (i=0; i<srt->nvset; i++)
+    {
+        fprintf(stderr,"dbg_vset %d:", i);
+        for (j=0; j<srt->vset[i].mask->n; j++) fprintf(stderr,"%c%lu",j==0?' ':':',srt->vset[i].mask->b[j]);
+        fprintf(stderr,"\t");
+        for (j=0; j<srt->vset[i].nvar; j++)
+        {
+            var_t *var = &srt->var[srt->vset[i].var[j]];
+            fprintf(stderr,"\t%s",var->str);
+            for (k=0; k<var->nvcf; k++)
+                fprintf(stderr,"%c%d", k==0?':':',',var->vcf[k]);
+        }
+        fprintf(stderr,"\n");
+    }
+}
+void debug_vbuf(sr_sort_t *srt)
+{
+    int i, j;
+    for (j=0; j<srt->vcf_buf[0].nrec; j++)
+    {
+        fprintf(stderr,"dbg_vbuf %d:\t", j);
+        for (i=0; i<srt->sr->nreaders; i++)
+        {
+            vcf_buf_t *buf = &srt->vcf_buf[i];
+            fprintf(stderr,"\t%d", buf->rec[j] ? buf->rec[j]->pos+1 : 0);
+        }
+        fprintf(stderr,"\n");
+    }
+}
+char *grp_create_key(sr_sort_t *srt)
+{
+    if ( !srt->str.l ) return strdup("");
+    int i;
+    hts_expand(char*,srt->noff,srt->mcharp,srt->charp);
+    for (i=0; i<srt->noff; i++)
+    {
+        srt->charp[i] = srt->str.s + srt->off[i];
+        if ( i>0 ) srt->charp[i][-1] = 0;
+    }
+    qsort(srt->charp, srt->noff, sizeof(*srt->charp), cmpstringp);
+    char *ret = (char*) malloc(srt->str.l + 1), *ptr = ret;
```

Comment:
Unchecked `malloc`!
I would be very happy if everything that could allocate memory checked to see if the memory was actually allocated.  This will need to be fixed at some point.